src/vmm/src/devices/virtio/net/metrics.rs

```diff
@@ -0,0 +1,300 @@
+    pub rx_fails: SharedIncMetric,
+    /// Number of successful read operations while receiving data.
+    pub rx_count: SharedIncMetric,
+    /// Number of times reading from TAP failed.
+    pub tap_read_fails: SharedIncMetric,
+    /// Number of times writing to TAP failed.
+    pub tap_write_fails: SharedIncMetric,
+    /// Number of transmitted bytes.
+    pub tx_bytes_count: SharedIncMetric,
+    /// Number of malformed TX frames.
+    pub tx_malformed_frames: SharedIncMetric,
+    /// Number of errors while transmitting data.
+    pub tx_fails: SharedIncMetric,
+    /// Number of successful write operations while transmitting data.
+    pub tx_count: SharedIncMetric,
+    /// Number of transmitted packets.
+    pub tx_packets_count: SharedIncMetric,
+    /// Number of TX partial reads from guest.
+    pub tx_partial_reads: SharedIncMetric,
+    /// Number of events associated with the transmitting queue.
+    pub tx_queue_event_count: SharedIncMetric,
+    /// Number of events associated with the rate limiter installed on the transmitting path.
+    pub tx_rate_limiter_event_count: SharedIncMetric,
+    /// Number of RX rate limiter throttling events.
+    pub tx_rate_limiter_throttled: SharedIncMetric,
+    /// Number of packets with a spoofed mac, sent by the guest.
+    pub tx_spoofed_mac_count: SharedIncMetric,
+}
+
+impl NetDeviceMetrics {
+    /// Const default construction.
+    pub const fn new() -> Self {
+        Self {
+            activate_fails: SharedIncMetric::new(),
+            cfg_fails: SharedIncMetric::new(),
+            mac_address_updates: SharedIncMetric::new(),
+            no_rx_avail_buffer: SharedIncMetric::new(),
+            no_tx_avail_buffer: SharedIncMetric::new(),
+            event_fails: SharedIncMetric::new(),
+            rx_queue_event_count: SharedIncMetric::new(),
+            rx_event_rate_limiter_count: SharedIncMetric::new(),
+            rx_partial_writes: SharedIncMetric::new(),
+            rx_rate_limiter_throttled: SharedIncMetric::new(),
+            rx_tap_event_count: SharedIncMetric::new(),
+            rx_bytes_count: SharedIncMetric::new(),
+            rx_packets_count: SharedIncMetric::new(),
+            rx_fails: SharedIncMetric::new(),
+            rx_count: SharedIncMetric::new(),
+            tap_read_fails: SharedIncMetric::new(),
+            tap_write_fails: SharedIncMetric::new(),
+            tx_bytes_count: SharedIncMetric::new(),
+            tx_malformed_frames: SharedIncMetric::new(),
+            tx_fails: SharedIncMetric::new(),
+            tx_count: SharedIncMetric::new(),
+            tx_packets_count: SharedIncMetric::new(),
+            tx_partial_reads: SharedIncMetric::new(),
+            tx_queue_event_count: SharedIncMetric::new(),
+            tx_rate_limiter_event_count: SharedIncMetric::new(),
+            tx_rate_limiter_throttled: SharedIncMetric::new(),
+            tx_spoofed_mac_count: SharedIncMetric::new(),
+        }
+    }
```

Comment:
Initially I used new() to be consistent with how other devices were initialised but then I think if default always provides a 0 initialised Atomicusize then we can just use that and avoid this.

Reply:
Yeah, all integer types in Rust have a default implementation that produces `0`. I think we can eliminate a bunch of explicit `new` functions this way actually, if we really don't need the const-ness :thinking: (although those are probably out of scope here)

Reply:
ok so METRICS is static and so FirecrackerMetrics::new() had to be a const which made the rest of structs define a const new and so NetDeviceMetrics had a new() instead of default. But since we moved out of logger and don't use NetDeviceMetrics directly in a static initialization we can remove it and just use default. Once we are done with all the device related metrics we can revisit the const for FirecrackerMetrics as well.

Reply:
What I meant is that you need neither `new` nor `default`, because in `FirecrackerMetrics` you can just use a struct initializer directly. Since its a unit struct, it can just be
```rs
impl FirecrackerMetrics {
    /// Const default construction.
    pub const fn new() -> Self {
        Self {
            ...
            net: NetDeviceMetrics {}, // or NetDeviceMetric,
            ...
        }
    }
}
```
definitely more of a nit though :)